executor/local/stage.go

```diff
@@ -6,26 +6,169 @@ package local
 
 import (
 	"context"
+	"fmt"
+	"strings"
+	"time"
 
+	"github.com/go-vela/types/constants"
 	"github.com/go-vela/types/pipeline"
 )
 
 // CreateStage prepares the stage for execution.
 func (c *client) CreateStage(ctx context.Context, s *pipeline.Stage) error {
+	// load the logs for the init step from the client
+	l, err := c.loadStepLogs(c.pipeline.Stages[0].Steps[0].ID)
+	if err != nil {
+		return err
+	}
+
+	// update the init log with progress
+	//
+	// https://pkg.go.dev/github.com/go-vela/types/library?tab=doc#Log.AppendData
+	l.AppendData([]byte(fmt.Sprintf("> Pulling step images for stage %s...\n", s.Name)))
+
+	// create the steps for the stage
+	for _, step := range s.Steps {
+		// TODO: make this not hardcoded
+		// update the init log with progress
+		//
+		// https://pkg.go.dev/github.com/go-vela/types/library?tab=doc#Log.AppendData
+		l.AppendData([]byte(fmt.Sprintf("$ docker image inspect %s\n", step.Image)))
+
+		// create the step
+		err := c.CreateStep(ctx, step)
+		if err != nil {
+			return err
+		}
+
+		// inspect the step image
+		image, err := c.Runtime.InspectImage(ctx, step)
+		if err != nil {
+			return err
+		}
+
+		// update the init log with step image info
+		//
+		// https://pkg.go.dev/github.com/go-vela/types/library?tab=doc#Log.AppendData
+		l.AppendData(image)
+	}
+
 	return nil
 }
 
 // PlanStage prepares the stage for execution.
 func (c *client) PlanStage(ctx context.Context, s *pipeline.Stage, m map[string]chan error) error {
+	// ensure dependent stages have completed
+	for _, needs := range s.Needs {
+		// check if a dependency stage has completed
+		stageErr, ok := m[needs]
+		if !ok { // stage not found so we continue
+			continue
+		}
+
+		// wait for the stage channel to close
+		select {
```

Comment:
I hadn't seen this before! [It](https://gobyexample.com/select) seems pretty powerful combined with [receivers](https://golang.org/ref/spec#Receive_operator).

Reply:
Yes you are correct 👍 

Typically I've only seen it used when trying to handle "infinite" loops or concurrency primitives in Go.

But it sure does come in handy when you need it 😅